### PR TITLE
Fix measuring remaining timestamp/duration metrics

### PR DESCRIPTION
**Pull request by serathius**
Fix missing conversions to milliseconds.
Fix measuring time by deferring anonymous function.
Convert metrics to milliseconds.

### CODE REVIEW

File: events/sources/kubernetes/kubernetes_source.go [hunk 1]
```diff
@@ -78,8 +78,12 @@ type KubernetesEventSource struct {
 
 func (this *KubernetesEventSource) GetNewEvents() *core.EventBatch {
 	startTime := time.Now()
-	defer lastEventTimestamp.Set(float64(time.Now().Unix()))
-	defer scrapEventsDuration.Observe(float64(time.Since(startTime)) / float64(time.Microsecond))
+	defer func() {
```
Comment (piosz) [inline events/sources/kubernetes/kubernetes_source.go]: @crassirostris do we want this change?

File: events/sources/kubernetes/kubernetes_source.go [hunk 2]
```diff
@@ -78,8 +78,12 @@ type KubernetesEventSource struct {
...
+		lastEventTimestamp.Set(float64(time.Now().Unix()))
+	}()
+	defer func() {
+		scrapEventsDuration.Observe(float64(time.Since(startTime)) / float64(time.Microsecond))
```
Comment (piosz) [inline events/sources/kubernetes/kubernetes_source.go]: Should be Milliseconds

File: metrics/sources/kubelet/kubelet.go [hunk 3]
```diff
@@ -260,7 +260,9 @@ func (this *kubeletMetricsSource) ScrapeMetrics(start, end time.Time) (*DataBatc
 
 func (this *kubeletMetricsSource) scrapeKubelet(client *KubeletClient, host Host, start, end time.Time) ([]cadvisor.ContainerInfo, error) {
 	startTime := time.Now()
-	defer kubeletRequestLatency.WithLabelValues(this.hostname).Observe(float64(time.Since(startTime)))
+	defer func() {
+		kubeletRequestLatency.WithLabelValues(this.hostname).Observe(float64(time.Since(startTime)) / float64(time.Microsecond))
```
Comment (piosz) [inline metrics/sources/kubelet/kubelet.go]: Milliseconds here and in the definition of the metrics.

File: metrics/sources/manager.go [hunk 4]
```diff
@@ -164,12 +164,16 @@ responseloop:
 func scrape(s MetricsSource, start, end time.Time) (*DataBatch, error) {
 	sourceName := s.Name()
 	startTime := time.Now()
-	defer lastScrapeTimestamp.
-		WithLabelValues(sourceName).
-		Set(float64(time.Now().Unix()))
-	defer scraperDuration.
-		WithLabelValues(sourceName).
-		Observe(float64(time.Since(startTime)) / float64(time.Microsecond))
+	defer func() {
```
Comment (piosz) [inline metrics/sources/manager.go]: Shouldn't be the begging of the scraping?

**Comment by crassirostris** [comment]
Please add some description to the PR

File: events/sources/kubernetes/kubernetes_source.go (hunk 1 above)
Comment (crassirostris) [reply events/sources/kubernetes/kubernetes_source.go]: I'm not sure

@serathius Could you please explain it somewhere, e.g. in the description?

Comment (serathius) [reply events/sources/kubernetes/kubernetes_source.go]: I was mainly motivated by https://github.com/kubernetes/heapster/pull/1643/files

File: metrics/sources/manager.go (hunk 4 above)
Comment (serathius) [reply metrics/sources/manager.go]: I wanted it to be consistent with other `last` metrics

File: events/sources/kubernetes/kubernetes_source.go (hunk 1 above)
Comment (crassirostris) [reply events/sources/kubernetes/kubernetes_source.go]: OK, I see, arguments evaluating

This looks ugly anyways, can it be refactored into a single function call?

Comment (serathius) [reply events/sources/kubernetes/kubernetes_source.go]: Yep

Comment (crassirostris) [reply events/sources/kubernetes/kubernetes_source.go]: OTOH, let it be, refactoring it seems to add even more complexity and boilerplate

**Comment by serathius** [comment]
done

File: metrics/sources/manager.go (hunk 4 above)
Comment (serathius) [reply metrics/sources/manager.go]: @crassirostris Should `last` event timestamp come from before or after execution?

**Comment by piosz** [comment]
Please agree with Mik, whether we want to report beginning or end of the operation.
Also squash commits before merge.

Otherwise LGTM.

File: metrics/sources/manager.go (hunk 4 above)
Comment (crassirostris) [reply metrics/sources/manager.go]: Making it consistent with `metrics/sources/manager.go` seems reasonable

Ambiguity of the description is another problem, but it's outside of this PR's scope

File: events/sources/kubernetes/kubernetes_source.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (serathius) [reply events/sources/kubernetes/kubernetes_source.go]: done

**Pull request merged**